src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLoggerProcessor.cs

```diff
@@ -31,63 +70,134 @@ public ConsoleLoggerProcessor(IConsole console, IConsole errorConsole)
                 Name = "Console logger queue processing thread"
             };
             _outputThread.Start();
+            s_meter.CreateObservableGauge<long>("queue-size", GetQueueSize);
         }
 
         public virtual void EnqueueMessage(LogMessageEntry message)
         {
-            if (!_messageQueue.IsAddingCompleted)
-            {
-                try
-                {
-                    _messageQueue.Add(message);
-                    return;
-                }
-                catch (InvalidOperationException) { }
-            }
-
-            // Adding is completed so just log the message
-            try
+            // cannot enqueue when adding is completed
+            if (!Enqueue(message))
             {
                 WriteMessage(message);
             }
-            catch (Exception) { }
         }
 
         // for testing
         internal void WriteMessage(LogMessageEntry entry)
         {
-            IConsole console = entry.LogAsError ? ErrorConsole : Console;
-            console.Write(entry.Message);
+            try
+            {
+                var messagesDropped = Interlocked.Exchange(ref _messagesDropped, 0);
+                if (messagesDropped != 0)
+                {
+                    System.Console.Error.WriteLine($"{messagesDropped} message(s) dropped because of queue size limit. Increase the queue size or decrease logging verbosity to avoid this.{Environment.NewLine}");
+                }
+
+                IConsole console = entry.LogAsError ? ErrorConsole : Console;
+                console.Write(entry.Message);
+            }
+            catch (Exception)
+            {
+                CompleteAdding();
+            }
         }
 
         private void ProcessLogQueue()
         {
-            try
+            while (!_isAddingCompleted || _messageQueue.Count > 0)
             {
-                foreach (LogMessageEntry message in _messageQueue.GetConsumingEnumerable())
+                if (TryDequeue(out LogMessageEntry message))
                 {
                     WriteMessage(message);
                 }
             }
-            catch
+        }
+
+        public bool Enqueue(LogMessageEntry item)
+        {
+            lock (_messageQueue)
+            {
+                while (_messageQueue.Count >= MaxQueueLength && !_isAddingCompleted)
+                {
+                    if (FullMode == ConsoleLoggerBufferFullMode.DropWrite)
+                    {
+                        Interlocked.Increment(ref _messagesDropped);
```

Comment:
If we change the dropped message handling via my suggestion above then we shouldn't need interlocked operations for this field any longer.